src/regex.cr

```diff
@@ -530,37 +530,81 @@ class Regex
     ret >= 0
   end
 
-  # Returns a `Hash` where the values are the names of capture groups and the
-  # keys are their indexes. Non-named capture groups will not have entries in
-  # the `Hash`. Capture groups are indexed starting from `1`.
+  # Yields the name of each capture group, along with the key as second argument. Non-named
+  # capture groups will not be yielded. Capture groups are indexed starting from `1`.
   #
   # ```
-  # /(.)/.name_table                         # => {}
-  # /(?<foo>.)/.name_table                   # => {1 => "foo"}
-  # /(?<foo>.)(?<bar>.)/.name_table          # => {2 => "bar", 1 => "foo"}
-  # /(.)(?<foo>.)(.)(?<bar>.)(.)/.name_table # => {4 => "bar", 2 => "foo"}
+  # /(?<foo>.)/.each_name do |name, index|
+  #   name  # => "foo"
+  #   index # => 1
+  # end
   # ```
-  def name_table : Hash(UInt16, String)
+  def each_name(& : String, UInt16 ->) : Nil
     LibPCRE.full_info(@re, @extra, LibPCRE::INFO_NAMECOUNT, out name_count)
     LibPCRE.full_info(@re, @extra, LibPCRE::INFO_NAMEENTRYSIZE, out name_entry_size)
-    table_pointer = Pointer(UInt8).null
-    LibPCRE.full_info(@re, @extra, LibPCRE::INFO_NAMETABLE, pointerof(table_pointer).as(Pointer(Int32)))
-    name_table = table_pointer.to_slice(name_entry_size*name_count)
 
-    lookup = Hash(UInt16, String).new
+    table_pointer = Pointer(UInt8).null
+    LibPCRE.full_info(@re, @extra, LibPCRE::INFO_NAMETABLE,
+      pointerof(table_pointer).as(Pointer(Int32)))
+    name_table = table_pointer.to_slice(name_entry_size * name_count)
 
     name_count.times do |i|
       capture_offset = i * name_entry_size
-      capture_number = (name_table[capture_offset].to_u16 << 8) | name_table[capture_offset + 1].to_u16
+      capture_number =
+        (name_table[capture_offset].to_u16 << 8) | name_table[capture_offset + 1].to_u16
 
       name_offset = capture_offset + 2
       checked = name_table[name_offset, name_entry_size - 3]
       name = String.new(checked.to_unsafe)
 
-      lookup[capture_number] = name
+      yield name, capture_number
+    end
+  end
+
+  # Returns a `Hash` where the values are the names of capture groups and the
+  # keys are their indexes. Non-named capture groups will not have entries in
+  # the `Hash`. Capture groups are indexed starting from `1`.
+  #
+  # ```
+  # /(.)/.name_table                         # => {}
+  # /(?<foo>.)/.name_table                   # => {1 => "foo"}
+  # /(?<foo>.)(?<bar>.)/.name_table          # => {2 => "bar", 1 => "foo"}
+  # /(.)(?<foo>.)(.)(?<bar>.)(.)/.name_table # => {4 => "bar", 2 => "foo"}
+  # ```
+  def name_table : Hash(UInt16, String)
+    Hash(UInt16, String).new.tap do |name_table|
+      each_name do |name, capture_number|
+        name_table[capture_number] = name
+      end
+    end
+  end
```

Comment:
The key type should most certainly not be `UInt16` but `Int32`.

I'm also wondering if an array would be a better data structure. With many unnamed groups it would occupy more memory, but lookup is more performant. And looking up items by index is a classic job for `Indexable`.

Reply:
IIRC you can't use the `Array` since the index keys might not be corresponded to array indexes.

Reply:
I just realized that this method already exists in the current API with this return type. So we shouldn't break it.
But nevertheless, it's bad. We shouldn't expose `UInt16`. That's just an implementation detail.
So I'd suggest to deprecate that method and add another one with a different name and correct return type.

Yes, in an array the indeces of unnamed capture groups must be nil values.

Reply:
Example:

```cr
/(.)(?<foo>.)(.)(?<bar>.)(.)/.name_table # => {4 => "bar", 2 => "foo"}
/(.)(?<foo>.)(.)(?<bar>.)(.)/.name_array # => [nil, "foo", nil, "bar"]
```